automation/run.ts

```diff
@@ -24,7 +24,16 @@ import {
 	ROOT,
 	runUnderMsys,
 } from './build-bin';
-import { release } from './deploy-bin';
+import {
+	release,
+	updateDescriptionOfReleasesAffectedByIssue1359,
+} from './deploy-bin';
+
+function exitWithError(error: Error | string): never {
+	console.error(`Error: ${error}`);
+	process.exit(1);
```

Comment:
Your comment about satisfying the compiler made me lookup the [docs for `process.exit()`](https://nodejs.org/dist/latest-v10.x/docs/api/process.html#process_process_exit_code) and I bumped on something interesting (that at least I didn't know):

> Calling process.exit() will force the process to exit as quickly as possible even if there are still asynchronous operations pending that have not yet completed fully, including I/O operations to process.stdout and process.stderr.

Based on that, you could theoretically get in a situation where you asked Node to log the error before exiting but it will never be printed to output. In fact, it calls out this exact code as a counter-example. The manual suggests you set `process.exitCode` instead and let it die naturally when the event loop has no more work to do, or just throw an error.

> Rather than calling process.exit() directly, the code should set the process.exitCode and allow the process to exit naturally by avoiding scheduling any additional work for the event loop. If it is necessary to terminate the Node.js process due to an error condition, throwing an uncaught error and allowing the process to terminate accordingly is safer than calling process.exit().

I've never seen this happening in practice, and it doesn't *really* matter for this script, but I wonder if we're using this pattern in the CLI which is expected to run in a variety of platforms, in which case I think it makes sense to err on the safe side and replace any `process.exit()`s with exceptions that are caught by an unhandled exception/rejection handler to cleanly print an error (instead of printing stacktraces as it happens by default) and let the process exit naturally.

Reply:
Indeed. One reason why the CLI doesn't terminate the process with an uncaught error is that uncaught errors are actually caught and reported via Sentry/Raven. Errors that we don't want to be reported via Sentry/Raven are usually "handled" by calling `exitWithExpectedError`, which calls `process.exit()`. One actual side effect of calling `exitWithExpectedError` that I am aware of, related to pending async operations, is that the mixpanel reporting is indeed interrupted before normal/graceful completion. But it's one of those things that "have been working well enough for years," changing the code would not fix any reported bug and might actually introduce new bugs and hurt performance a little bit...